### PR TITLE
Add p2p-preheat policy UI

### LINKED ISSUE #12283: [UI]: Create page for preheat policy management

### LINKED ISSUE #12284: [UI]: Create page for viewing the preheating execution history (task executions)

### LINKED ISSUE #12451: Adjust input field size in P2P UI
In this popup dialog,  all input fields are too narrow. Please increase the size/width of these components.

Screen Shot 2020-07-08 at 4 29 32 PM

### PR DESCRIPTION
**Pull request by AllForNothing**
Fixes #12451 
Fixes #12284
Fixes #12283
Temp Env: https://10.185.12.35
UX: https://www.figma.com/proto/cPY9Lh8LbIQawAJSGmi1KA/Harbor-v2.0?node-id=1390%3A1&scaling=min-zoom
Changed points:
image
image

Signed-off-by: AllForNothing <sshijun@vmware.com>

### CODE REVIEW

File: src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts [hunk 1]
```diff
@@ -196,6 +208,14 @@ export class DistributionSetupModalComponent implements OnInit {
       if ( this.model.endpoint !== this.originModelForEdit.endpoint) {
         return true;
       }
+      // tslint:disable-next-line:triple-equals
+      if ( this.model.enabled != this.originModelForEdit.enabled) {
+        return true;
+      }
+      // tslint:disable-next-line:triple-equals
+      if ( this.model.insecure != this.originModelForEdit.insecure) {
```
Comment (jwangyangls) [inline src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts:216]: Why donot use '!=='  and delete  '// tslint:disable-next-line:triple-equals'

File: src/portal/src/app/project/p2p-provider/add-p2p-policy/add-p2p-policy.component.ts [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    }
+    // tslint:disable-next-line:triple-equals
+    if (this.originTagsForEdit != this.tags) {
+      return true;
+    }
+    // tslint:disable-next-line:triple-equals
+    if (this.originOnlySignedImagesForEdit != this.onlySignedImages) {
+      return true;
+    }
+    // tslint:disable-next-line:triple-equals
```
Comment (jwangyangls) [inline src/portal/src/app/project/p2p-provider/add-p2p-policy/add-p2p-policy.component.ts:262]: above

File: src/portal/src/app/project/p2p-provider/p2p-provider.service.ts [hunk 3]
```diff
@@ -0,0 +1,103 @@
+// Copyright Project Harbor Authors
```
Comment (jwangyangls) [inline src/portal/src/app/project/p2p-provider/p2p-provider.service.ts:1]: The content has only some declared methods or const, and no service is used, so there is no need to create a service.

File: src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+      + `/projects/${this.projectName}/preheat/policies/${this.preheatPolicyName}/executions/${this.executionId}/tasks/${taskId}/logs`;
+  }
+  clrLoadTasks(withLoading): void {
+      if (withLoading) {
+        this.loading = true;
+      }
+      this.preheatService.ListTasks({
+        projectName: this.projectName,
+        preheatPolicyName: this.preheatPolicyName,
+        executionId: +this.executionId
```
Comment (jwangyangls) [inline src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts:171]: Does it support backend paging?

**Comment by selina-cui** [comment]
Add find one bug in your test env. Steps to reproduce
1. Go Preheat tab in project details page.
2. Select one policy, the related tasks can display.
3. Open browser developer tool, select another policy, watch the link sent in network inspector, the policy name in the request is not correct.

File: src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts (hunk 1 above)
Comment (AllForNothing) [reply src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts:216]: "", null and undefined need to be regarded as same value

File: src/portal/src/app/project/p2p-provider/p2p-provider.service.ts (hunk 3 above)
Comment (AllForNothing) [reply src/portal/src/app/project/p2p-provider/p2p-provider.service.ts:1]: There same public methords

File: src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts (hunk 4 above)
Comment (AllForNothing) [reply src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts:171]: for task list, no!

File: src/portal/src/app/project/p2p-provider/add-p2p-policy/add-p2p-policy.component.html [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+            <clr-checkbox-wrapper>
+              <input clrCheckbox type="checkbox" id="onlySignedImages" [(ngModel)]="onlySignedImages" name="onlySignedImages">
+              <label for="onlySignedImages">{{'P2P_PROVIDER.ONLY_SIGNED' | translate}}</label>
+            </clr-checkbox-wrapper>
+          </clr-checkbox-container>
+          <div class="clr-form-control mt-0" *ngIf="withNotary() && enableContentTrust && !onlySignedImages">
+            <label for="repo" class="width-6rem"></label>
+            <div class="clr-control-container">
+              <label class="label label-warning  warning">
+                <span>{{'P2P_PROVIDER.CONTENT_WARNING' | translate}}</span>
```
[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/p2p-provider/add-p2p-policy/add-p2p-policy.component.html:99]: > Add find one bug in your test env. Steps to reproduce
> 
> 1. Go Preheat tab in project details page.
> 2. Select one policy, the related tasks can display.
> 3. Open browser developer tool, select another policy, watch the link sent in network inspector, the policy name in the request is not correct.

It's surely a bug. Fixed

[the remark this answers is not in the document]
Comment (AllForNothing) [reply src/portal/src/app/project/p2p-provider/add-p2p-policy/add-p2p-policy.component.html:99]: After development, will ask Alex and Louis to check UI appearance。If needed, will improve

File: src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts (hunk 4 above)
Comment (jwangyangls) [reply src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts:171]: not support?

File: src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts (hunk 1 above)
Comment (jwangyangls) [reply src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts:216]: I don’t think he will have undefined value. If it does, it means there are other errors.

Comment (jwangyangls) [reply src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts:216]: I don’t think he will have undefined value. If it does, it means there are other errors.

File: src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts (hunk 4 above)
Comment (AllForNothing) [reply src/portal/src/app/project/p2p-provider/task-list/task-list.component.ts:171]: image

File: src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts (hunk 1 above)
Comment (AllForNothing) [reply src/portal/src/app/distribution/distribution-setup-modal/distribution-setup-modal.component.ts:216]: Supposed that i created a  policy by call api   and   did not set insecure property,  when i edit this policy , and set its insecure to false.    so its original value is 'undefined'  and current value is 'false'. if use '!==', we get false, but infact , we should regard these two value  are equal

**Pull request merged**